stablehlo/reference/ProcessGrid.cpp

```diff
@@ -0,0 +1,128 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "stablehlo/reference/ProcessGrid.h"
+
+#include <condition_variable>
+#include <mutex>
+#include <utility>
+
+#include "llvm/Support/ErrorHandling.h"
+#include "stablehlo/reference/Tensor.h"
+
+namespace mlir {
+namespace stablehlo {
+
+SmallVector<ProcessGroup> ProcessGroups::find(ProcessId &processId) {
+  SmallVector<ProcessGroup> groupsFound{};
+
+  for (auto processGroup : *this)
+    for (auto id : processGroup)
+      if (id == processId) groupsFound.push_back(processGroup);
+
+  return groupsFound;
+}
+
+ProcessGroups Process::crossReplica(
+    SmallVector<SmallVector<uint32_t>> replicaGroups) {
+  return grid->crossReplica(replicaGroups);
+}
+
+ProcessGroups Process::crossPartition(
+    SmallVector<SmallVector<uint32_t>> partitionGroups) {
+  return grid->crossPartition(partitionGroups);
+}
+
+SmallVector<std::pair<ProcessId, Tensor>> Process::rendezvous(
+    ProcessGroup processGroup, int64_t channelId, const Tensor &operand) {
+  return grid->rendezvous(processGroup, channelId, id, operand);
+}
+
+ProcessGrid::ProcessGrid(uint32_t numReplicas, uint32_t numPartitions)
+    : numReplicas_(numReplicas),
+      numPartitions_(numPartitions),
+      resourceLockMap_(
+          std::map<std::pair<ProcessGroup, int64_t>, std::mutex>()),
+      resourceConditionMap_(std::map<std::pair<ProcessGroup, int64_t>,
+                                     std::condition_variable>()),
+      channels_(std::map<std::pair<ProcessGroup, int64_t>,
+                         SmallVector<std::pair<ProcessId, Tensor>>>()) {}
+
+ProcessGroups ProcessGrid::crossReplica(
+    SmallVector<SmallVector<uint32_t>> replicaGroups) {
+  ProcessGroups processGroups;
+  for (auto replicaGroup : replicaGroups) {
+    for (uint32_t partitionId = 0; partitionId < numPartitions_;
+         ++partitionId) {
+      ProcessGroup processGroup;
+      for (uint32_t replicaId : replicaGroup)
+        processGroup.push_back({replicaId, partitionId});
+      processGroups.push_back(processGroup);
+    }
+  }
+  return processGroups;
+}
+
+ProcessGroups ProcessGrid::crossPartition(
+    SmallVector<SmallVector<uint32_t>> partitionGroups) {
+  ProcessGroups processGroups;
+  for (auto partitionGroup : partitionGroups) {
+    for (uint32_t replicaId = 0; replicaId < numReplicas_; ++replicaId) {
+      ProcessGroup processGroup;
+      for (uint32_t partitionId : partitionGroup)
+        processGroup.push_back({replicaId, partitionId});
+      processGroups.push_back(processGroup);
+    }
+  }
+  return processGroups;
+}
+
+SmallVector<std::pair<ProcessId, Tensor>> ProcessGrid::rendezvous(
+    ProcessGroup processGroup, int64_t channelId, ProcessId &processId,
+    const Tensor &operand) {
+  {
+    std::unique_lock<std::mutex> lk(
+        resourceLockMap_[{processGroup, channelId}]);
```

Comment:
From https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly: "Avoid abbreviations unless they are well known". I don't think we're gaining much by saving two characters at the cost of reducing readability.